File: src/api/phenoage.ts
/** PhenoAge API client. The Levine formula stays on the server. */

export const PHENOAGE_DISCLAIMER =
  'PhenoAge is a research index (Levine et al., Aging 2018), not a medical service, diagnosis, or treatment recommendation.';

export interface PhenoAgeScore {
  chronologicalAge: number;
  phenoAge: number;
  ageDelta: number;
  mortalityScore10yr: number;
  disclaimer: string;
}

interface PhenoAgePayload {
  chronological_age: number;
  pheno_age: number;
  age_delta: number;
  mortality_score_10yr: number;
  disclaimer: string;
}

function apiUrl(path: string): string {
  const base = import.meta.env.VITE_API_BASE_URL ?? '';
  return `${base}${path}`;
}

export async function fetchPhenoAge(
  chronologicalAge: number,
  biomarkers: Record<string, number>,
  signal?: AbortSignal,
): Promise<PhenoAgeScore> {
  const response = await fetch(apiUrl('/api/v1/phenoage'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      chronological_age: chronologicalAge,
      markers: {
        albumin: biomarkers.albumin,
        creatinine: biomarkers.creatinine,
        glucose: biomarkers.glucose,
        crp: biomarkers.crp,
        lymphocyte: biomarkers.lymphocyte,
        mcv: biomarkers.mcv,
        rdw: biomarkers.rdw,
        alp: biomarkers.alp,
        wbc: biomarkers.wbc,
      },
    }),
    signal,
  });
  if (!response.ok) {
    throw new Error(`PhenoAge request failed (${response.status})`);
  }
  const payload = (await response.json()) as PhenoAgePayload;
  return {
    chronologicalAge: payload.chronological_age,
    phenoAge: payload.pheno_age,
    ageDelta: payload.age_delta,
    mortalityScore10yr: payload.mortality_score_10yr,
    disclaimer: payload.disclaimer,
  };
}

File: app/api/phenoage.py
"""PhenoAge HTTP surface. The Levine formula lives in the service, not here."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, status

from app.domain.schemas import PhenoAgeRequest, PhenoAgeResponse
from app.services.phenoage import (
    CanonicalBiomarkers,
    PhenoAgeInputError,
    calculate_canonical_phenoage,
)

router = APIRouter(prefix="/api/v1/phenoage", tags=["phenoage"])


@router.post("", response_model=PhenoAgeResponse)
def score_phenoage(body: PhenoAgeRequest) -> PhenoAgeResponse:
    """Score nine biomarkers. Nothing is stored."""
    markers = body.markers
    try:
        result = calculate_canonical_phenoage(
            CanonicalBiomarkers(
                albumin_g_l=markers.albumin,
                creatinine_mg_dl=markers.creatinine,
                glucose_mg_dl=markers.glucose,
                crp_mg_l=markers.crp,
                lymphocyte_percent=markers.lymphocyte,
                mcv_fl=markers.mcv,
                rdw_percent=markers.rdw,
                alp_u_l=markers.alp,
                wbc_10e3_per_ul=markers.wbc,
                age_years=body.chronological_age,
            )
        )
    except PhenoAgeInputError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="PhenoAge inputs are not usable",
        ) from exc
    return PhenoAgeResponse(
        chronological_age=result.chronological_age,
        pheno_age=result.pheno_age,
        age_delta=result.age_delta,
        mortality_score_10yr=result.mortality_score_10yr,
        disclaimer=result.disclaimer,
    )
